src/twisted/protocols/_smb/core.py

```diff
@@ -0,0 +1,519 @@
+            if self.dialect > MAX_DIALECT:
+                raise base.SMBError(
+                    "min client dialect %04x higher than our max %04x" %
+                    (self.dialect, MAX_DIALECT))
+            log.debug("dialect {dlt:04x} chosen", dlt=self.dialect)
+        resp = COMMANDS[0][2]()
+        resp.size = 65
+        resp.signing = NEGOTIATE_SIGNING_ENABLED
+        resp.dialect = self.dialect
+        resp.server_uuid = self.factory.server_uuid.bytes_le
+        resp.capabilities = GLOBAL_CAP_DFS
+        resp.max_transact = MAX_TRANSACT_SIZE
+        resp.max_read = MAX_READ_SIZE
+        resp.max_write = MAX_WRITE_SIZE
+        resp.time = base.u2nt_time(time.time())
+        resp.boot_time = base.u2nt_time(self.factory.server_start)
+        resp.offset = 128
+        resp.buflen = len(blob)
+        resp.buffer = blob
+        self.send_with_header(resp, 'negotiate')
+
+    def smb_session_setup(self, req, resp):
+        blob = req.buffer[req.offset - len(req) - 64:req.offset - len(req) -
+                          64 + req.buflen]
+        log.debug("SESSION SETUP")
+        log.debug("-------------")
+        log.debug("Size             {sz}", sz=req.size)
+        log.debug("Security mode    0x{sm:08x}", sm=req.security_mode)
+        log.debug("Capabilities     0x{cap:08x}", cap=req.capabilities)
+        log.debug("Channel          0x{chl:08x}", chl=req.channel)
+        log.debug("Prev. session ID 0x{pid:016x}", pid=req.prev_session_id)
+        if self.first_session_setup:
+            self.blob_manager.receiveInitialBlob(blob)
+            resp.buffer = self.blob_manager.generateChallengeBlob()
+            self.session_setup_response(resp, STATUS_MORE_PROCESSING)
+            self.first_session_setup = False
+        else:
+            self.blob_manager.receiveResp(blob)
+            if self.blob_manager.credential:
+                log.debug("got credential: %r" % self.blob_manager.credential)
+                d = self.factory.portal.login(
+                    self.blob_manager.credential,
+                    SMBMind(req.prev_session_id,
+                            self.blob_manager.credential.domain, self.addr),
+                    ISMBServer)
+                d.addCallback(self._cb_login, resp)
+                d.addErrback(self._eb_login, resp)
+            else:
+                resp.buffer = self.blob_manager.generateChallengeBlob()
+                self.session_setup_response(resp, STATUS_MORE_PROCESSING)
+
+    def _cb_login(self, t, resp):
+        _, self.avatar, self.logout_thunk = t
+        resp.buffer = self.blob_manager.generateAuthResponseBlob(True)
+        log.debug("successful login")
+        self.session_setup_response(resp, STATUS_SUCCESS)
+
+    def _eb_login(self, failure, resp):
+        log.debug(failure.getTraceback())
+        resp.buffer = self.blob_manager.generateAuthResponseBlob(False)
+        self.session_setup_response(resp, STATUS_LOGON_FAILURE)
+
```

Comment:
I think it might be better to define these inline rather than as private methods; private methods that are only supposed to be called as specific callbacks in specific places just pollute the class namespace and add confusion for future maintainers about when & where they're OK to call.